File: solutions/ruby/dnd-character/1/dnd_character.rb
class DndCharacter
  attr_reader :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints
  
  def self.modifier(score)
    ((score - 10) / 2).floor
  end

  def initialize
    @strength = ability
    @dexterity = ability
    @constitution = ability
    @intelligence = ability
    @wisdom = ability
    @charisma = ability
    @hitpoints = 10 + DndCharacter.modifier(@constitution)
  end

  def ability
    Array.new(4) { rand(1..6) }
      .sort
      .last(3)
      .sum
  end
end

File: solutions/c/dnd-character/1/dnd_character.c
#include "dnd_character.h"
#include <stdlib.h>
#include <math.h>

int ability(void) {
   int dados[4];
   int menor = 7;
   int soma_total = 0;

   for (int i = 0; i < 4; i++) {
      dados[i] = (rand() % 6) + 1;
      soma_total += dados[i];
      
      if (dados[i] < menor) {
         menor = dados[i];
      }
   }
   return soma_total - menor;
}

int modifier(int score) {
   return (int)floor((score - 10) / 2.0);
}


dnd_character_t make_dnd_character(void) {
   dnd_character_t character;

   character.strength = ability();
   character.dexterity = ability();
   character.constitution = ability();
   character.intelligence = ability();
   character.wisdom = ability();
   character.charisma = ability();

   character.hitpoints = 10 + modifier(character.constitution);

   return character;
}